From 77f63786efb6b345a07d6339dfba838cc5225f7b Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 11 Oct 2023 00:36:07 +0530
Subject: [PATCH] hostapd: handle link_id in nl80211 radar events

With Multi-Link Operation (MLO), there is requirement to handle the link_id
with nl80211 radar related events.

This includes -
 * passing correct link_id during start radar event
 * using link_id to identify the correct link_bss in order to route the event

Add changes for the same.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 src/ap/ap_drv_ops.c                | 10 +++++++
 src/drivers/driver_nl80211.c       | 15 +++++++++-
 src/drivers/driver_nl80211_event.c | 44 ++++++++++++++++++++++++------
 3 files changed, 59 insertions(+), 10 deletions(-)

diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 37761a4d9995..f53395a14f7c 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -935,6 +935,16 @@ int hostapd_start_dfs_cac(struct hostapd_iface *iface,
 	}
 	data.radar_background = radar_background;
 
+	data.link_id = -1;
+
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap) {
+		data.link_id = hapd->mld_link_id;
+		wpa_printf(MSG_DEBUG,
+			   "hostapd_start_dfs_cac: link_id=%d", data.link_id);
+	}
+#endif /* CONFIG_IEEE80211BE */
+
 	res = hapd->driver->start_dfs_cac(hapd->drv_priv, &data);
 	if (!res) {
 		if (radar_background)
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index fe802bde207f..2da66e164076 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10466,9 +10466,22 @@ static int nl80211_start_radar_detection(void *priv,
 		return -1;
 	}
 
+	if (nl80211_link_valid(bss, freq->link_id)) {
+		wpa_printf(MSG_DEBUG, "nl80211: radar detection (CAC) on link_id=%d",
+			   freq->link_id);
+
+		if (nla_put_u8(msg, NL80211_ATTR_MLO_LINK_ID, freq->link_id)) {
+			nlmsg_free(msg);
+			return -1;
+		}
+	}
+
 	ret = send_and_recv_msgs(drv, msg, NULL, NULL, NULL, NULL);
-	if (ret == 0)
+	if (ret == 0) {
+		nl80211_link_set_freq(bss, freq->link_id, freq->freq);
 		return 0;
+	}
+
 	wpa_printf(MSG_DEBUG, "nl80211: Failed to start radar detection: "
 		   "%d (%s)", ret, strerror(-ret));
 	return -1;
diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 61491efa523a..608eb93247f1 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -2392,11 +2392,13 @@ static void nl80211_connect_failed_event(struct wpa_driver_nl80211_data *drv,
 }
 
 
-static void nl80211_radar_event(struct wpa_driver_nl80211_data *drv,
-				struct nlattr **tb)
+static void nl80211_radar_event(struct i802_bss *bss, struct nlattr **tb)
 {
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct i802_link *mld_link;
 	union wpa_event_data data;
 	enum nl80211_radar_event event_type;
+	void *ctx = bss->ctx;
 
 	if (!tb[NL80211_ATTR_WIPHY_FREQ] || !tb[NL80211_ATTR_RADAR_EVENT])
 		return;
@@ -2440,25 +2442,49 @@ static void nl80211_radar_event(struct wpa_driver_nl80211_data *drv,
 		   data.dfs_event.chan_offset, data.dfs_event.chan_width,
 		   data.dfs_event.cf1, data.dfs_event.cf2);
 
+	mld_link = nl80211_get_mld_link_by_freq(bss, data.dfs_event.freq);
+	if (!mld_link) {
+		/* For non-ML case also, %NULL would be returned. Hence check if
+		 * its for non-ML case, then proceed with first link.
+		 * Non-ML link should have link ID as -1 and freq should be set.
+		 */
+		if (bss->n_links == 1 && bss->links[0].link_id == -1 &&
+		    bss->links[0].freq) {
+			mld_link = bss->flink;
+			goto process_radar_event;
+		}
+
+		wpa_printf(MSG_DEBUG, "nl80211: DFS event on unknown freq on %s",
+			   bss->ifname);
+		return;
+	}
+
+process_radar_event:
+	if (mld_link->link_id != NL80211_DRV_LINK_ID_NA) {
+		wpa_printf(MSG_DEBUG, "nl80211: DFS event for link_id %d",
+			   mld_link->link_id);
+		ctx = mld_link->ctx;
+	}
+
 	switch (event_type) {
 	case NL80211_RADAR_DETECTED:
-		wpa_supplicant_event(drv->ctx, EVENT_DFS_RADAR_DETECTED, &data);
+		wpa_supplicant_event(ctx, EVENT_DFS_RADAR_DETECTED, &data);
 		break;
 	case NL80211_RADAR_CAC_FINISHED:
-		wpa_supplicant_event(drv->ctx, EVENT_DFS_CAC_FINISHED, &data);
+		wpa_supplicant_event(ctx, EVENT_DFS_CAC_FINISHED, &data);
 		break;
 	case NL80211_RADAR_CAC_ABORTED:
-		wpa_supplicant_event(drv->ctx, EVENT_DFS_CAC_ABORTED, &data);
+		wpa_supplicant_event(ctx, EVENT_DFS_CAC_ABORTED, &data);
 		break;
 	case NL80211_RADAR_NOP_FINISHED:
-		wpa_supplicant_event(drv->ctx, EVENT_DFS_NOP_FINISHED, &data);
+		wpa_supplicant_event(ctx, EVENT_DFS_NOP_FINISHED, &data);
 		break;
 	case NL80211_RADAR_PRE_CAC_EXPIRED:
-		wpa_supplicant_event(drv->ctx, EVENT_DFS_PRE_CAC_EXPIRED,
+		wpa_supplicant_event(ctx, EVENT_DFS_PRE_CAC_EXPIRED,
 				     &data);
 		break;
 	case NL80211_RADAR_CAC_STARTED:
-		wpa_supplicant_event(drv->ctx, EVENT_DFS_CAC_STARTED, &data);
+		wpa_supplicant_event(ctx, EVENT_DFS_CAC_STARTED, &data);
 		break;
 	default:
 		wpa_printf(MSG_DEBUG, "nl80211: Unknown radar event %d "
@@ -3994,7 +4020,7 @@ static void do_process_drv_event(struct i802_bss *bss, int cmd,
 		mlme_event_ft_event(drv, tb);
 		break;
 	case NL80211_CMD_RADAR_DETECT:
-		nl80211_radar_event(drv, tb);
+		nl80211_radar_event(bss, tb);
 		break;
 	case NL80211_CMD_STOP_AP:
 		nl80211_stop_ap(drv, tb);
-- 
2.17.1

